### PR TITLE
Migrated favorites

### LINKED ISSUE #4267: Opening favorites results in "unable to open folder xxx that already exists but is not a directory"
**Describe the bug**

When my colleague updated from ZE 3.4.2 to 3.5.0, she got these errors when trying to open her favorites:

Image

And it fails to open the favorites profile as a result.

These are standard PDS favorites and work fine in 3.4.2.

**To Reproduce**

I can't reproduce 🙁
 
She has about 40 data sets in her favorites and doesn't want to do the workaround I'd suggest which is to remove the zosmf profile from her favorites and re-favorite those data sets, as it is just easier for her to downgrade to 3.4.2 for the time being until this gets fixed.

**Expected behavior**

It to open favorites.

**Screenshots**

**Desktop (please complete the following information):**

- OS: Win11
- Zowe Explorer Version: 3.5.0
- (Optional) Zowe CLI Version:
- (Optional) Do you have secure credentials enabled?
- (Optional) Are you using a custom credential manager?

**Additional context**

2 other colleagues now reported the same issue to me.

### PR DESCRIPTION
**Pull request by jace-roell**
## Proposed changes

When adding a data set to favorites in ZE and then migrating said data set. It would result in a broken state since there was no support favoriting migrated DS. 

This PR adds support for favoriting migrated data sets and handles various edge cases. 

**Scenarios to test**

With **PDS** or **Seq**
1. Recalling a favorited data set
2. Migrating a favorited data set
3. favoriting an already migrated data set
4. using an old version of ZE, favoriting a DS, migrating the DS, and switching to this branch to test the rendering

Note: A right-click option for "Refresh Data Set" has been added for migrated DS nodes

## Release Notes

Milestone:

Changelog:

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**All checks must pass before this pull request is reviewed by the Zowe Explorer squad.**

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [ ] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`
- [ ] New ZE APIs are tested with extender types that haven't adopted yet to determine breaking changes. Can use Zowe zFTP marketplace extension.

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have tested new functionality with the FTP extension and profile verifying no extender profile type breakages introduced
- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

### CODE REVIEW

File: packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts [hunk 1]
```diff
@@ -360,6 +414,7 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
                         contextOverride: Constants.DS_MIGRATED_FILE_CONTEXT,
                         profile: cachedProfile,
                     });
+                    dsNode.wasPds = item.dsorg?.startsWith("PO") ? true : undefined;
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts]: Is the ternary required? `wasPds` is a boolean, so `undefined` and `false` would have the same impact when used in logical expressions

```suggestion
                    dsNode.wasPds = item.dsorg?.startsWith("PO");
```

File: packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts [hunk 2]
```diff
@@ -512,11 +567,12 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
             }
 
             if (SharedContext.isSession(this)) {
-                const dsTree = SharedTreeProviders.ds as DatasetTree;
-                // Reset and remove previous search patterns in case pattern has changed
-                dsTree.resetFilterForChildren(this.children);
-                // set new search patterns for each child of getChildren
-                dsTree.applyPatternsToChildren(this.children, this.patternMatches);
+                if (dsTree) {
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts]: Consider joining this check with `SharedContext.isSession(this)` in the outer `if` statement to reduce cognitive complexity, e.g. `if (SharedContext.isSession(this) && dsTree) { /* .. */ }`

Rationale: if `dsTree` is falsy, then none of the code within this `if` statement runs - so the assertions can be joined into one `if` condition

File: packages/zowe-explorer/src/trees/dataset/DatasetFSProvider.ts [hunk 3]
```diff
@@ -429,13 +430,21 @@ export class DatasetFSProvider extends BaseProvider implements vscode.FileSystem
...
             }
         });
 
+        if (isMigrated && entryExists && entry?.type === vscode.FileType.Directory) {
+            const parentDir = this.lookupParentDirectory(uri);
+            const dsname = uriPath[Number(pdsMember)];
+            parentDir.entries.delete(dsname);
+            entryExists = false;
+        }
+
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/dataset/DatasetFSProvider.ts:439]: Leaving a comment here since lines 448-452 were not changed in this PR... but I'm curious to know if we would benefit from a `if (!isMigrated)` check around the `fetchEntriesForDataset` call?

Rationale: We already use the migration status in the new `if` statement on L441-447. This prevents us from fetching members of the data set if we know it is migrated.

**Review by traeok — COMMENTED** [review]
Thanks for your work on this Jace! I left a couple suggestions around data set node logic, and had a question around optimizing the behavior of the `fetchDataset` function.

**Review by anaxceron — APPROVED** [review]

File: packages/zowe-explorer/src/tools/ZowePersistentFilters.ts [hunk 4]
```diff
@@ -25,6 +25,7 @@ export class ZowePersistentFilters {
     private static readonly favorites: string = "favorites";
     private static readonly vsamFavorites: string = "vsamFavorites";
     private static readonly memberFavorites: string = "memberFavorites";
+    private static readonly migratedFavorites: string = "migratedFavorites";
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/tools/ZowePersistentFilters.ts:28]: Is the addition of this persistent setting necessary? I don't see `migratedFavorites` being written to anywhere, only loaded in `readFavorites` and cleared in `updateFavorites`.

File: packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts [hunk 5]
```diff
@@ -236,10 +245,12 @@ export class ZoweDatasetNode extends ZoweTreeNode implements IZoweDatasetTreeNod
 
         // Create entry in the filesystem to represent the recalled data set
         if (isPds) {
+            DatasetFSProvider.instance.createDirectory(this.resourceUri);
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts:248]: Curious why we need to invoke `createDirectory` twice for a PDS, both via our own FSP instance and through the VSCode FS API?

File: packages/zowe-explorer/src/trees/dataset/DatasetFSProvider.ts [hunk 6]
```diff
@@ -1143,4 +1152,19 @@ export class DatasetFSProvider extends BaseProvider implements vscode.FileSystem
...
     }
+
+    public invalidateCache(uri: vscode.Uri): void {
+        try {
+            const parent = this.lookupParentDirectory(uri, true);
+            if (parent) {
+                const basename = path.posix.basename(uri.path);
+                if (parent.entries.has(basename)) {
+                    parent.entries.delete(basename);
+                    this.fireSoon({ type: vscode.FileChangeType.Deleted, uri });
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/trees/dataset/DatasetFSProvider.ts]: Since I believe we try to fire `FileChangeType` events only when a corresponding change has happened remotely, am curious why we fire a `Deleted` event when the DS hasn't necessarily been deleted. Is the intent to propagate invalidation to any cache that extenders may be keeping?

File: packages/zowe-explorer/src/trees/dataset/DatasetTree.ts [hunk 7]
```diff
@@ -1487,8 +1553,11 @@ Would you like to do this now?`,
                         memberFavoritesArray.push(favoriteEntry);
                     }
                 } else {
-                    const favoriteEntry =
-                        "[" + profileNode.label.toString() + "]: " + favorite.label.toString() + "{" + SharedContext.getBaseContext(favorite) + "}";
+                    let baseContext = SharedContext.getBaseContext(favorite);
+                    if (SharedContext.isMigrated(favorite)) {
+                        baseContext = pdsNode.wasPds ? "pds_migr" : "ds_migr";
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/trees/dataset/DatasetTree.ts]: If the base context is already `pds` or `ds`, can we simplify this by removing the ternary?
```suggestion
                        baseContext += "_migr";
```

My assumption is based on setting a breakpoint here to check the value of `baseContext`, but may be invalid if there are other possible values for base context of a DS node.

If it is valid, then I'm curious if `wasPds` could be removed altogether - besides this place in the code it only seems to be used as a fallback value for determining `isPds` but in most cases I'd expect `dsorg` to be present for this.

**Review by JWaters02 — APPROVED** [review]
Functionality wise, all the issues I was seeing previously are now resolved. I'll re-review if any logic changes come out of timothy's reviews. Thanks

File: packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts (hunk 5 above)
Comment (jace-roell) [reply packages/zowe-explorer/src/trees/dataset/ZoweDatasetNode.ts:248]: I think the diff view betrayed me. Removed the extra call 😅

**Review by traeok — APPROVED** [review]
LGTM, thanks Jace for the fix!

File: packages/zowe-explorer/src/trees/dataset/DatasetTree.ts [hunk 8]
```diff
@@ -1487,8 +1553,11 @@ Would you like to do this now?`,
                         memberFavoritesArray.push(favoriteEntry);
                     }
                 } else {
-                    const favoriteEntry =
-                        "[" + profileNode.label.toString() + "]: " + favorite.label.toString() + "{" + SharedContext.getBaseContext(favorite) + "}";
+                    let baseContext = SharedContext.getBaseContext(favorite);
+                    if (SharedContext.isMigrated(favorite)) {
+                        baseContext += Constants.CONTEXT_PREFIX + Constants.DS_MIGRATED_FILE_CONTEXT;
+                    }
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/trees/dataset/DatasetTree.ts]: Should this change be reverted? It seems that if a dataset was already migrated, then `getBaseContext` will return `migr`. The new logic repeats it twice: `migr_migr`

**Review by t1m0thyj — CHANGES_REQUESTED** [review]

**Comment by JWaters02** [comment]
Functionality wise everything still seems to be working as expected for me from the changes since my approval

File: packages/zowe-explorer/src/trees/dataset/DatasetTree.ts [hunk 9]
```diff
@@ -579,7 +579,8 @@ export class DatasetTree extends ZoweTreeProvider<IZoweDatasetTreeNode> implemen
         const lines: string[] = this.mPersistence.readFavorites();
         const vsamLines: string[] = this.mPersistence.readVsamFavorites();
         const memberLines: string[] = this.mPersistence.readMemberFavorites();
-        const combinedLines = [...lines, ...vsamLines, ...memberLines];
+        const migratedLines: string[] = this.mPersistence.readMigratedFavorites();
+        const combinedLines = [...lines, ...vsamLines, ...memberLines, ...migratedLines];
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/dataset/DatasetTree.ts]: (nitpick) Consider joining declaration and assignment

```
const combinedLines = [
    ...this.mPersistence.readFavorites(),
    ...this.mPersistence.readVsamFavorites(),
    ...this.mPersistence.readMemberFavorites(),
    ...this.mPersistence.readMigratedFavorites()
];
```

File: packages/zowe-explorer/src/trees/dataset/DatasetActions.ts [hunk 10]
```diff
@@ -2176,6 +2176,46 @@ export class DatasetActions {
...
+     * @param node The migrated dataset node to refresh
+     */
+    private static async refreshMigratedDataSet(node: IZoweDatasetTreeNode): Promise<void> {
+        const dataSetName = node.label as string;
+        try {
+            ZoweLogger.info(`Refreshing migrated data set ${dataSetName}`);
+            const statusMsg = Gui.setStatusBarMessage(`$(sync~spin) ${vscode.l10n.t("Checking migration status...")}`);
+            const mvsApi = ZoweExplorerApiRegister.getMvsApi(node.getProfile());
+            const resp = await mvsApi.dataSet(dataSetName, { attributes: true });
+            statusMsg.dispose();
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/dataset/DatasetActions.ts]: If the `dataSet` API fails with an error, the status bar message is not disposed. Consider declaring `statusMsg` outside the try/catch, assigning it, and then disposing it in event of an error

**Review by traeok — APPROVED** [review]

**Review by t1m0thyj — APPROVED** [review]
Thanks @jace-roell, tested with the FTP extension and favorites seem to be working as expected for extenders 👍
